File: code/algo/getoffer/go/3/trie_tree_count.go
package main

type TrieNode struct {
	children [26]*TrieNode
	val      int
}

type MapSum struct {
	root *TrieNode
	cnt  map[string]int
}

func NewMapSum() MapSum {
	return MapSum{
		&TrieNode{},
		map[string]int{},
	}
}

func (m *MapSum) Insert(key string, val int) {
	delta := val
	if m.cnt[key] > 0 {
		delta -= m.cnt[key]
	}
	m.cnt[key] = val
	node := m.root
	for _, ch := range key {
		ch -= 'a'
		if node.children[ch] == nil {
			node.children[ch] = &TrieNode{}
		}
		node = node.children[ch]
		node.val += delta
	}
}

func (m *MapSum) Sum(prefix string) int {
	node := m.root
	for _, ch := range prefix {
		ch -= 'a'
		if node.children[ch] == nil {
			return 0
		}
		node = node.children[ch]
	}
	return node.val
}

File: code/algo/getoffer/c++/3/trie_tree_count.cpp
#include <string>
#include <unordered_map>
#include <vector>

class TrieNode {
public:
  TrieNode *children[26] = {nullptr};
  int val = 0;
};

class MapSum {
public:
  MapSum() { root = new TrieNode(); }

  void Insert(const std::string &key, int val) {
    int delta = val;
    if (cnt.find(key) != cnt.end()) {
      delta -= cnt[key];
    }
    cnt[key] = val;
    TrieNode *node = root;
    for (char ch : key) {
      ch -= 'a';
      if (node->children[ch] == nullptr) {
        node->children[ch] = new TrieNode();
      }
      node = node->children[ch];
      node->val += delta;
    }
  }

  int Sum(const std::string &prefix) {
    TrieNode *node = root;
    for (char ch : prefix) {
      ch -= 'a';
      if (node->children[ch] == nullptr) {
        return 0;
      }
      node = node->children[ch];
    }
    return node->val;
  }

private:
  TrieNode *root;
  std::unordered_map<std::string, int> cnt;
};
